### PR TITLE
Spw ranges fourier filter

### PR DESCRIPTION
**Pull request by aewallwi**
PR Enables Fourier Filtering of multiple subbands simultaneously while keeping track of edge flags between subbands.

This should be merged before #700 so we can add the `filter_spw_ranges` arg to `tophat_frfilter_run.py` in #700.

### CODE REVIEW

**Review by jsdillon — CHANGES_REQUESTED** [review]
Some minor changes, but otherwise looks fine.

One general comment I have is that this seems rather complicated for what it's trying to do, which I think will make it hard to maintain. It's not clear to me if it can be simplified without a significant loss of functionality, so I leave it to you @aewallwi to determine if it's sufficiently necessary to warrant the extra complexity.

File: hera_cal/vis_clean.py [hunk 1]
```diff
@@ -21,6 +21,40 @@
 from .flag_utils import factorize_flags
 
 
+def find_discontinuity_edges(x, xtol=1e-3):
+    """Find edges based on discontinuity in x-axis
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:25]: I've added comments to explain this but basically, this function helps us find discontinuities in the x-axis which exist if we have several non-contiguous subbands in our data (for example, excising the FM band).

File: hera_cal/vis_clean.py [hunk 2]
```diff
@@ -21,6 +21,40 @@
...
+
+    Parameters
+    ----------
+    x: array-like
+        1d numpy array of x-values.
+    xtol: float, optional
+        fractional discontinuity in diff of xs above which an edge will be identified.
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:32]: I've clarified this.

Median would work as long as >50% of the channels are in fact contiguous.

File: hera_cal/vis_clean.py [hunk 3]
```diff
@@ -43,50 +77,122 @@ def truncate_flagged_edges(data_in, weights_in, x, ax='freq'):
...
+        x with completely flagged edges trimmed off.
     dout : array-like 2d.
-            data_in with completely flagged edges trimmed off.
+        data_in with completely flagged edges trimmed off.
     wout : array-like 2d.
-            weights_in with completely flagged edges trimmed off.
-    edges : 2-list of 2-tuples
-            the width of the edges trimmed.
+        weights_in with completely flagged edges trimmed off.
+    edges : list of 2-tuples or 2-list of lists of 2-tuples
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:85]: I've added an example that hopefully helps.

File: hera_cal/vis_clean.py [hunk 4]
```diff
@@ -43,50 +77,122 @@ def truncate_flagged_edges(data_in, weights_in, x, ax='freq'):
...
+            ind_right = chunk[1] - chunk[0]
+            chunkslice = slice(chunk[0], chunk[1])
+            unflagged_chans = np.where(~np.all(np.isclose(weights_in[:, chunkslice], 0.0), axis=0))[0]
+            if np.count_nonzero(unflagged_chans) > 0:
+                # truncate data to be filtered where appropriate.
+                ind_left = np.min(unflagged_chans)
+                ind_right = np.max(unflagged_chans) + 1
+            inds_left.append(ind_left)
+            inds_right.append(ind_right)
+        if len(inds_left) > 1:
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:130]: i think your right. I've gotten rid of the conditional.

File: hera_cal/vis_clean.py [hunk 5]
```diff
@@ -105,17 +212,30 @@ def flag_rows_with_flags_within_edge_distance(weights_in, min_flag_edge_distance
 
     """
     if ax == 'time':
-        wout = flag_rows_with_flags_within_edge_distance(weights_in.T, min_flag_edge_distance).T
+        wout = flag_rows_with_flags_within_edge_distance(x, weights_in.T, min_flag_edge_distance).T
     else:
+        if isinstance(x, (tuple, list)) and len(x) == 2:
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:228]: I don't see how? I think if x is a list of length 2 this conditional will eval as True?

[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:228]: OK, I think I see what you mean... x is a list of doubles.

[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:228]: I've modified this line to make sure the elements of x are ndarrays, lists, or tuples as well.

File: hera_cal/vis_clean.py [hunk 6]
```diff
@@ -184,12 +304,12 @@ def get_max_contiguous_flag_from_filter_periods(x, filter_centers, filter_half_w
         to the largest delay in the filter centers and filter_widths
     """
     if len(x) == 2:
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py]: See response above.

File: hera_cal/vis_clean.py [hunk 7]
```diff
@@ -184,12 +304,12 @@ def get_max_contiguous_flag_from_filter_periods(x, filter_centers, filter_half_w
...
-        dx = [np.mean(np.diff(x[0])), np.mean(np.diff(x[1]))]
+        dx = [np.median(np.diff(x[0])), np.median(np.diff(x[1]))]
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py]: I've added an exception to catch this

File: hera_cal/vis_clean.py [hunk 8]
```diff
@@ -872,7 +994,10 @@ def fourier_filter(self, filter_centers, filter_half_widths, mode='clean',
                 x = self.freqs
         else:
             raise ValueError("ax must be one of ['freq', 'time', 'both']")
-
+        if filter_spw_ranges is None:
+            filter_spw_ranges = [(0, self.Nfreqs)]
+        if len(self.freqs) != np.sum([spw_range[1] - spw_range[0] for spw_range in filter_spw_ranges]) or not np.allclose(self.freqs, self.freqs[np.hstack([np.arange(spw_range[0], spw_range[1]).astype(int) for spw_range in filter_spw_ranges])]):
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:1010]: I've broken this one up.

File: hera_cal/vis_clean.py [hunk 9]
```diff
@@ -872,7 +994,10 @@ def fourier_filter(self, filter_centers, filter_half_widths, mode='clean',
...
+            raise NotImplementedError("No support for spw-ranges that do not disjointly cover the entire frequency band.")
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:1011]: This is checking whether all of the channels in the original frequency axis are contained in one of the user specified SPWs. I've made the error state this more clearly.

File: hera_cal/vis_clean.py [hunk 10]
```diff
@@ -907,119 +1032,158 @@ def fourier_filter(self, filter_centers, filter_half_widths, mode='clean',
...
+                # just in case any artifacts make it through after our other flagging rounds
+                # flag integrations or channels where the RMS of the model exceeds the RMS of the unflagged data
+                # by some threshold.
+                if flag_model_rms_outliers:
+                    skipped = flag_model_rms(skipped, d, w, mdl, model_rms_threshold=model_rms_threshold, ax=ax)
+
+                # also flag skipped edge channels and integrations.
+                if skip_flagged_edges:
+                    if ax == 'both':
+                        print(chunks[0], edges[0])
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/vis_clean.py:1156]: Thanks for catching this. These should have been deleted.

**Comment by aewallwi** [comment]
Thanks for the review @jsdillon! This is ready for another look whenever you get a chance.

**Review by jsdillon — APPROVED** [review]
Looks good to me!

**Pull request merged**